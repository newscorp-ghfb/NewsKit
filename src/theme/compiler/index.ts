/* eslint-disable @typescript-eslint/no-explicit-any */
import {Theme, UncompiledTheme} from '../types';
import {ThemeCompilerOptions} from './types';
import {generateInkStylePresets} from './ink-style-preset-generator';
import {recurseUnknown} from '../../utils/recurse-unknown';

export const compileTheme = (
  theme: UncompiledTheme | Theme,
  options: ThemeCompilerOptions = {},
): Theme => {
  if (!theme || theme.compiled === true) {
    return theme;
  }

  // eslint-disable-next-line no-console
  const errorLogger = options.errorLogger || console.error.bind(console);

  const uncompiledTheme = {
    ...theme,
    stylePresets: {
      ...theme.stylePresets,
      ...generateInkStylePresets(theme),
    },
  };

  const {icons = {}, ...filteredTheme} = uncompiledTheme;
  return {
    ...(recurseUnknown(uncompiledTheme, filteredTheme, errorLogger) as any),
    icons,
    compiled: true,
  };
};
