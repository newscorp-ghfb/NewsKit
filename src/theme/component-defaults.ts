export const componentDefaults = {
  assistiveText: {
    small: {
      stylePreset: 'assistiveText',
      typographyPreset: 'utilityBody010',
      minHeight: 'sizing030',
      startEnhancer: {
        iconSize: 'iconSize010',
        spaceInline: 'space020',
      },
      endEnhancer: {
        iconSize: 'iconSize010',
        spaceInline: 'space020',
      },
    },
    medium: {
      stylePreset: 'assistiveText',
      typographyPreset: 'utilityBody020',
      minHeight: 'sizing030',
      startEnhancer: {
        iconSize: 'iconSize020',
        spaceInline: 'space020',
      },
      endEnhancer: {
        iconSize: 'iconSize020',
        spaceInline: 'space020',
      },
    },
    large: {
      stylePreset: 'assistiveText',
      typographyPreset: 'utilityBody030',
      minHeight: 'sizing030',
      startEnhancer: {
        iconSize: 'iconSize020',
        spaceInline: 'space030',
      },
      endEnhancer: {
        iconSize: 'iconSize020',
        spaceInline: 'space030',
      },
    },
  },
  audioPlayer: {
    seekBar: {
      slider: {
        track: {
          stylePreset: 'audioPlayerSeekBarTrack',
          size: 'sizing020',
        },
        indicator: {
          stylePreset: 'audioPlayerSeekBarIndicator',
        },
        thumb: {
          stylePreset: 'audioPlayerThumb',
          size: 'sizing040',
        },
        thumbLabel: {
          stylePreset: 'audioPlayerLabels',
        },
        labels: {
          stylePreset: 'audioPlayerLabels',
          typographyPreset: 'utilityLabel020',
        },
      },
      buffering: {
        stylePreset: 'audioPlayerSeekBarBuffering',
      },
    },
    controls: {
      space: 'space030',
      previousButton: {
        stylePreset: 'iconButtonMinimalPrimary',
      },
      replayButton: {
        stylePreset: 'iconButtonMinimalPrimary',
      },
      playPauseButton: {
        stylePreset: 'iconButtonSolidPrimary',
      },
      forwardButton: {
        stylePreset: 'iconButtonMinimalPrimary',
      },
      nextButton: {
        stylePreset: 'iconButtonMinimalPrimary',
      },
      popoutButton: {
        stylePreset: 'iconButtonMinimalPrimary',
      },
    },
    volumeControl: {
      slider: {
        track: {
          stylePreset: 'volumeControlTrack',
          size: 'sizing010',
        },
        indicator: {
          stylePreset: 'volumeControlIndicator',
        },
        thumb: {
          stylePreset: 'audioPlayerThumb',
          size: 'sizing040',
        },
        thumbLabel: {
          stylePreset: 'volumeControlLabels',
        },
        labels: {
          stylePreset: 'iconButtonMinimalPrimary',
        },
      },
    },
  },
  byline: {
    stylePreset: 'inkSubtle',
    typographyPreset: 'utilityMeta020',
    spaceStack: 'space020',
    link: {
      stylePreset: 'linkInline',
      typographyPreset: 'utilityMeta020',
    },
    divider: {
      stylePreset: 'inkNonEssential',
      spaceInline: 'space020',
    },
  },
  caption: {
    typographyPreset: 'editorialCaption010',
    stylePreset: 'inkBase',
    spaceStack: 'space040',
    credit: {
      typographyPreset: 'utilityMeta010',
      stylePreset: 'uppercaseInkSubtle',
    },
  },
  captionInset: {
    typographyPreset: 'editorialCaption010',
    stylePreset: 'inkBase',
    spaceStack: 'space040',
    spaceInset: {
      xs: 'spaceInset040',
      md: 'spaceInset050',
    },
    credit: {
      typographyPreset: 'utilityMeta010',
      stylePreset: 'uppercaseInkSubtle',
    },
  },
  card: {
    stylePreset: 'cardContainer',
    horizontalRatio: '1:1',
    mediaContainer: {
      stylePreset: 'cardContainerMedia',
      spaceInline: {
        xs: 'space040',
        sm: 'space040',
        md: 'space050',
        lg: 'space050',
      },
    },
    teaserContainer: {
      stylePreset: 'cardContainerTeaser',
      spaceInset: 'spaceInsetSquish000',
    },
    actionsContainer: {
      stylePreset: 'cardContainerActions',
      minHeight: 'sizing080',
      spaceInset: 'spaceInsetSquish000',
    },
    headline: {
      nonInteractive: {
        typographyPreset: {
          xs: 'editorialHeadline020',
          sm: 'editorialHeadline020',
          md: 'editorialHeadline030',
          lg: 'editorialHeadline050',
        },
        heading: {
          stylePreset: 'inkContrast',
        },
        kicker: {
          stylePreset: 'uppercaseInkBrand010',
          spaceInline: {
            xs: 'space020',
            sm: 'space020',
            md: 'space030',
            lg: 'space030',
          },
        },
      },
      interactive: {
        typographyPreset: {
          xs: 'editorialHeadline020',
          sm: 'editorialHeadline020',
          md: 'editorialHeadline030',
          lg: 'editorialHeadline050',
        },
        heading: {
          stylePreset: 'headlineHeadingInteractive',
        },
        kicker: {
          stylePreset: 'headlineKickerInteractive',
          spaceInline: {
            xs: 'space020',
            sm: 'space020',
            md: 'space030',
            lg: 'space030',
          },
        },
      },
    },
  },
  cardInset: {
    stylePreset: 'cardContainer',
    horizontalRatio: '1:1',
    mediaContainer: {
      stylePreset: 'cardContainerMedia',
      spaceInline: 'space000',
    },
    teaserContainer: {
      stylePreset: 'cardContainerTeaser',
      spaceInset: {
        xs: 'spaceInset040',
        sm: 'spaceInset040',
        md: 'spaceInset050',
        lg: 'spaceInset050',
      },
    },
    actionsContainer: {
      stylePreset: 'cardContainerActions',
      minHeight: 'sizing000',
      spaceInset: {
        xs: 'spaceInsetSquish030',
        sm: 'spaceInsetSquish030',
        md: 'spaceInsetSquish040',
        lg: 'spaceInsetSquish040',
      },
    },
  },
  dateTime: {
    stylePreset: 'inkSubtle',
    typographyPreset: 'utilityMeta020',
    prefix: {
      stylePreset: 'inkSubtle',
      typographyPreset: 'utilityMeta020',
    },
    suffix: {
      stylePreset: 'inkSubtle',
      typographyPreset: 'utilityMeta020',
    },
  },
  drawer: {
    overlay: {
      zIndex: 70,
      transitionPreset: 'fade',
    },
    panel: {
      zIndex: 80,
      stylePreset: 'drawerPanel',
      maxSize: '100%',
      minSize: '20px',
      size: {
        xs: '305px',
        sm: '309px',
        md: '310px',
        lg: '333px',
        xl: '354px',
      },
      left: {
        transitionPreset: ['fade', 'slideLeft'],
      },
      right: {
        transitionPreset: ['fade', 'slideRight'],
      },
      top: {
        transitionPreset: ['fade', 'slideTop'],
      },
      bottom: {
        transitionPreset: ['fade', 'slideBottom'],
      },
    },
    header: {
      spaceInset: 'spaceInsetSquish040',
    },
    content: {
      spaceInset: 'spaceInset050',
    },
    closeButton: {
      stylePreset: 'iconButtonMinimalSecondary',
      spaceInset: 'spaceInset020',
    },
  },
  grid: {
    maxWidth: '1920px',
    columns: 12,
    containerMargin: {
      xs: 'space040',
      sm: 'space040',
      md: 'space040',
      lg: 'space050',
      xl: 'space050',
    },
    columnGutters: {
      xs: 'space040',
      sm: 'space040',
      md: 'space040',
      lg: 'space050',
      xl: 'space050',
    },
    rowGutters: {
      xs: 'space040',
      sm: 'space040',
      md: 'space040',
      lg: 'space050',
      xl: 'space050',
    },
  },
  headline: {
    typographyPreset: {
      xs: 'editorialHeadline060',
      sm: 'editorialHeadline060',
      md: 'editorialHeadline070',
      lg: 'editorialHeadline080',
    },
    heading: {
      stylePreset: 'inkContrast',
    },
    kicker: {
      stylePreset: 'uppercaseInkBrand010',
      spaceInline: {
        xs: 'space030',
        sm: 'space030',
        md: 'space040',
        lg: 'space040',
      },
    },
  },
  headlineH1: {
    stylePreset: 'inkContrast',
    typographyPreset: {
      xs: 'editorialHeadline050',
      lg: 'editorialHeadline070',
      xl: 'editorialHeadline080',
    },
  },
  headlineH2: {
    stylePreset: 'inkContrast',
    typographyPreset: {
      xs: 'editorialHeadline030',
      md: 'editorialHeadline040',
    },
  },
  headlineH3: {
    stylePreset: 'inkContrast',
    typographyPreset: {
      xs: 'editorialHeadline020',
      md: 'editorialHeadline030',
    },
  },
  headlineH4: {
    stylePreset: 'inkContrast',
    typographyPreset: {
      xs: 'editorialHeadline010',
      md: 'editorialHeadline020',
    },
  },
  headlineH5: {
    stylePreset: 'inkContrast',
    typographyPreset: 'editorialSubheadline020',
  },
  headlineH6: {
    stylePreset: 'inkContrast',
    typographyPreset: 'editorialSubheadline010',
  },
  icons: {
    stylePreset: 'iconDefault',
  },
  image: {
    stylePreset: 'imageSharp',
  },
  indeterminateProgressIndicator: {
    stylePreset: 'iconDefault',
    size: 'iconSize020',
  },
  inlineDrawer: {
    __extends: '{{componentDefaults.drawer}}',
    panel: {
      __extends: '{{componentDefaults.drawer.panel}}',
      maxSize: 'auto',
    },
  },
  inlineMessage: {
    stylePreset: 'inlineMessageInformative',
    spaceInset: 'spaceInset030',
    icon: {
      spaceInline: 'space030',
    },
    content: {
      title: {
        spaceStack: 'space030',
        stylePreset: 'inkContrast',
        typographyPreset: 'utilityHeading010',
      },
      message: {
        stylePreset: 'inkBase',
        typographyPreset: 'utilityBody020',
      },
    },
  },
  label: {
    small: {
      stylePreset: 'label',
      typographyPreset: 'utilityLabel010',
      spaceStack: 'space040',
      spaceInline: 'space010',
    },
    medium: {
      stylePreset: 'label',
      typographyPreset: 'utilityLabel020',
      spaceStack: 'space040',
      spaceInline: 'space010',
    },
    large: {
      stylePreset: 'label',
      typographyPreset: 'utilityLabel030',
      spaceStack: 'space040',
      spaceInline: 'space010',
    },
  },
  link: {
    stylePreset: 'linkInline',
    spaceInline: 'space010',
    externalIcon: {
      size: 'iconSize010',
    },
  },
  linkStandalone: {
    stylePreset: 'linkStandalone',
    typographyPreset: 'utilityLabel020',
  },
  modal: {
    overlay: {
      zIndex: 70,
      transitionPreset: {
        extend: 'fade',
        enterActive: {
          transitionDuration: '{{motions.motionDuration030}}',
          transitionTimingFunction: '{{motions.motionTimingEaseOut}}',
        },
        exitActive: {
          transitionDuration: '{{motions.motionDuration030}}',
          transitionDelay: '{{motions.motionDuration010}}',
          transitionTimingFunction: '{{motions.motionTimingEaseIn}}',
        },
      },
    },
    panel: {
      zIndex: 80,
      stylePreset: 'modalPanel',
      transitionPreset: [
        {
          extend: 'fade',
          enterActive: {
            transitionDuration: '{{motions.motionDuration010}}',
            transitionTimingFunction: '{{motions.motionTimingEaseOut}}',
            transitionDelay: '{{motions.motionDuration010}}',
          },
          exitActive: {
            transitionDuration: '{{motions.motionDuration010}}',
            transitionTimingFunction: '{{motions.motionTimingLinear}}',
          },
        },
        {
          extend: 'moveUp',
          enterActive: {
            transitionDuration: '{{motions.motionDuration010}}',
            transitionTimingFunction: '{{motions.motionTimingEaseOut}}',
            transitionDelay: '{{motions.motionDuration010}}',
          },
          exitActive: {
            transitionDuration: '{{motions.motionDuration010}}',
            transitionTimingFunction: '{{motions.motionTimingEaseIn}}',
          },
        },
      ],
      topOffset: '20vh',
      width: {
        xs: '90vw',
        sm: '60vw',
        md: '45vw',
        lg: '38vw',
        xl: '31vw',
      },
      minHeight: '15vh',
      maxHeight: {
        xs: '95vh',
        md: '80vh',
      },
    },
    header: {
      spaceInset: 'spaceInsetSquish040',
    },
    content: {
      spaceInset: 'spaceInset050',
    },
    closeButton: {
      stylePreset: 'iconButtonMinimalSecondary',
      spaceInset: 'spaceInset020',
    },
  },
  orderedList: {
    spaceInline: 'space040',
    content: {
      stylePreset: 'inkBase',
      typographyPreset: 'editorialParagraph010',
    },
    counter: {
      stylePreset: 'inkBase',
      typographyPreset: 'editorialParagraph010',
      minWidth: 'sizing050',
    },
  },
  overlay: {
    stylePreset: 'overlay',
    transitionPreset: 'fade',
  },
  paragraph: {
    stylePreset: 'inkBase',
    typographyPreset: 'editorialParagraph020',
    dropCap: {
      stylePreset: 'inkContrast',
      typographyPreset: {
        xs: 'editorialDropCap010',
        md: 'editorialDropCap020',
        lg: 'editorialDropCap030',
      },
      space: 'space010',
    },
  },
  scroll: {
    horizontal: {
      controls: {
        button: {
          stylePreset: 'iconButtonSolidPrimary',
          spaceInset: 'spaceInset000',
        },
        offset: 'space040',
      },
      overlays: {
        stylePreset: 'scrollOverlaysHorizontal',
        size: 'sizing090',
      },
    },
    vertical: {
      controls: {
        button: {
          stylePreset: 'iconButtonSolidPrimary',
          spaceInset: 'spaceInset000',
        },
        offset: 'space040',
      },
      overlays: {
        stylePreset: 'scrollOverlaysVertical',
        size: 'sizing090',
      },
    },
  },
  select: {
    small: {
      button: {
        stylePreset: 'inputField',
        minHeight: 'sizing060',
        width: '100%',
        spaceInset: 'spaceInset020',
        spaceStack: 'space020',
        spaceInline: 'space020',
        typographyPreset: 'utilityBody020',
        iconSize: 'iconSize020',
        loadingIndicator: {
          stylePreset: 'indeterminateProgressIndicatorPrimary',
        },
        startEnhancer: {
          spaceInline: 'space020',
          iconSize: 'iconSize020',
        },
        endEnhancer: {
          spaceInline: 'space020',
          iconSize: 'iconSize020',
        },
      },
      panel: {
        spaceStack: 'space010',
        stylePreset: 'selectPanel',
        maxHeight: '184px',
        spaceInset: 'spaceInset020',
      },
    },
    medium: {
      button: {
        stylePreset: 'inputField',
        minHeight: 'sizing080',
        width: '100%',
        spaceInset: 'spaceInset020',
        spaceStack: 'space020',
        spaceInline: 'space020',
        typographyPreset: 'utilityBody020',
        iconSize: 'iconSize020',
        loadingIndicator: {
          stylePreset: 'indeterminateProgressIndicatorPrimary',
        },
        startEnhancer: {
          spaceInline: 'space020',
          iconSize: 'iconSize020',
        },
        endEnhancer: {
          spaceInline: 'space020',
          iconSize: 'iconSize020',
        },
      },
      panel: {
        spaceStack: 'space010',
        stylePreset: 'selectPanel',
        maxHeight: '272px',
        spaceInset: 'spaceInset020',
      },
    },
    large: {
      button: {
        stylePreset: 'inputField',
        minHeight: 'sizing090',
        width: '100%',
        spaceInset: 'spaceInset030',
        spaceStack: 'space020',
        spaceInline: 'space020',
        typographyPreset: 'utilityBody030',
        iconSize: 'iconSize020',
        loadingIndicator: {
          stylePreset: 'indeterminateProgressIndicatorPrimary',
        },
        startEnhancer: {
          spaceInline: 'space020',
          iconSize: 'iconSize020',
        },
        endEnhancer: {
          spaceInline: 'space020',
          iconSize: 'iconSize020',
        },
      },
      panel: {
        spaceStack: 'space010',
        stylePreset: 'selectPanel',
        maxHeight: '360px',
        spaceInset: 'spaceInset020',
      },
    },
  },
  selectOption: {
    small: {
      minHeight: 'sizing060',
      spaceInline: 'space020',
      spaceInset: 'spaceInsetSquish010',
      stylePreset: 'selectOptionItem',
      typographyPreset: 'utilityBody020',
      icon: {
        stylePreset: 'selectOptionItemIcon',
        iconSize: 'iconSize020',
      },
    },
    medium: {
      minHeight: 'sizing080',
      spaceInline: 'space020',
      spaceInset: 'spaceInset020',
      stylePreset: 'selectOptionItem',
      typographyPreset: 'utilityBody020',
      icon: {
        stylePreset: 'selectOptionItemIcon',
        iconSize: 'iconSize020',
      },
    },
    large: {
      minHeight: 'sizing090',
      spaceInline: 'space020',
      spaceInset: 'spaceInsetStretch030',
      stylePreset: 'selectOptionItem',
      typographyPreset: 'utilityBody030',
      icon: {
        stylePreset: 'selectOptionItemIcon',
        iconSize: 'iconSize020',
      },
    },
  },
  shareBar: {
    label: {
      stylePreset: 'shareBarLabel',
      typographyPreset: 'utilityLabel020',
      spaceInline: 'space040',
    },
    items: {
      spaceInline: 'space020',
    },
  },
  slider: {
    track: {
      stylePreset: 'sliderTrack',
      size: 'sizing030',
    },
    indicator: {
      stylePreset: 'sliderIndicator',
    },
    thumb: {
      stylePreset: 'sliderThumb',
      size: 'sizing060',
    },
    thumbLabel: {
      stylePreset: 'sliderThumbLabel',
      typographyPreset: 'utilityLabel030',
      space: 'space060',
    },
    labels: {
      stylePreset: 'sliderLabels',
      typographyPreset: 'utilityLabel020',
      space: 'space030',
    },
  },
  standfirst: {
    styledText: {
      stylePreset: 'inkBase',
      typographyPreset: {
        xs: 'editorialSubheadline010',
        lg: 'editorialSubheadline020',
      },
    },
  },
  structuredList: {
    divider: {
      stylePreset: 'divider',
    },
  },
  structuredListItem: {
    stylePreset: 'structuredListItem',
    spaceInset: 'spaceInsetStretch040',
    minHeight: 'sizing100',
    icon: {
      size: 'iconSize010',
      stylePreset: 'inkContrast',
    },
  },

  textField: {
    small: {
      stylePreset: 'inputField',
      spaceInset: 'spaceInset020',
      minHeight: 'sizing060',
      width: '100%',
      typographyPreset: 'utilityBody020',
      spaceStack: 'space020',
      startEnhancer: {
        iconSize: 'iconSize020',
        spaceInline: 'space020',
      },
      endEnhancer: {
        iconSize: 'iconSize020',
        spaceInline: 'space020',
      },
    },
    medium: {
      stylePreset: 'inputField',
      spaceInset: 'spaceInset030',
      minHeight: 'sizing080',
      width: '100%',
      typographyPreset: 'utilityBody020',
      spaceStack: 'space020',
      startEnhancer: {
        iconSize: 'iconSize020',
        spaceInline: 'space020',
      },
      endEnhancer: {
        iconSize: 'iconSize020',
        spaceInline: 'space020',
      },
    },
    large: {
      stylePreset: 'inputField',
      spaceInset: 'spaceInset030',
      minHeight: 'sizing090',
      width: '100%',
      typographyPreset: 'utilityBody030',
      spaceStack: 'space020',
      startEnhancer: {
        iconSize: 'iconSize020',
        spaceInline: 'space020',
      },
      endEnhancer: {
        iconSize: 'iconSize020',
        spaceInline: 'space020',
      },
    },
  },
  textInput: {
    small: {
      label: {
        stylePreset: 'textInputLabel',
        typographyPreset: 'utilityLabel030',
        spaceStack: 'space040',
        spaceInline: 'space010',
      },
      input: {
        stylePreset: 'textInput',
        spaceInset: 'spaceInset020',
        minHeight: 'sizing060',
        typographyPreset: 'utilityBody020',
        spaceStack: 'space020',
        leadingIcon: {
          iconOffset: 'space030',
          spaceInset: 'space070',
        },
        validationIcon: {
          iconOffset: 'space030',
          spaceInset: 'space070',
          iconSize: 'iconSize020',
        },
      },
      assistiveText: {
        stylePreset: 'textInputAssistiveText',
        typographyPreset: 'utilityLabel020',
        minHeight: 'sizing030',
      },
    },
    medium: {
      label: {
        stylePreset: 'textInputLabel',
        typographyPreset: 'utilityLabel030',
        spaceStack: 'space040',
        spaceInline: 'space010',
      },
      input: {
        stylePreset: 'textInput',
        spaceInset: 'spaceInset030',
        minHeight: 'sizing080',
        typographyPreset: 'utilityBody020',
        spaceStack: 'space020',
        leadingIcon: {
          iconOffset: 'space030',
          spaceInset: 'space070',
        },
        validationIcon: {
          iconOffset: 'space030',
          spaceInset: 'space070',
          iconSize: 'iconSize020',
        },
      },
      assistiveText: {
        stylePreset: 'textInputAssistiveText',
        typographyPreset: 'utilityLabel020',
        minHeight: 'sizing030',
      },
    },
    large: {
      label: {
        stylePreset: 'textInputLabel',
        typographyPreset: 'utilityLabel030',
        spaceStack: 'space040',
        spaceInline: 'space010',
      },
      input: {
        stylePreset: 'textInput',
        spaceInset: 'spaceInset030',
        minHeight: 'sizing090',
        typographyPreset: 'utilityBody030',
        spaceStack: 'space020',
        leadingIcon: {
          iconOffset: 'space030',
          spaceInset: 'space080',
        },
        validationIcon: {
          iconOffset: 'space030',
          spaceInset: 'space080',
          iconSize: 'iconSize030',
        },
      },
      assistiveText: {
        stylePreset: 'textInputAssistiveText',
        typographyPreset: 'utilityLabel030',
        minHeight: 'sizing030',
      },
    },
  },
  titleBar: {
    stylePreset: 'titleBar',
    spaceInset: {
      xs: 'spaceInsetSquish030',
      lg: 'spaceInsetSquish040',
    },
    heading: {
      typographyPreset: {
        xs: 'editorialHeadline050',
        md: 'editorialHeadline070',
        lg: 'editorialHeadline080',
      },
    },
  },
  unorderedList: {
    spaceStack: 'space040',
    content: {
      stylePreset: 'inkBase',
      typographyPreset: 'editorialParagraph010',
    },
    marker: {
      stylePreset: 'inkBase',
      spaceInline: 'space020',
      size: 'iconSize005',
    },
  },
  volumeControl: {
    slider: {
      track: {
        stylePreset: 'volumeControlTrack',
      },
      indicator: {
        stylePreset: 'volumeControlIndicator',
      },
      thumb: {
        stylePreset: 'volumeControlThumb',
      },
    },
    button: {
      stylePreset: 'iconButtonMinimalPrimary',
      iconSize: 'iconSize020',
      size: 'small',
    },
  },
  checkbox: {
    small: {
      spaceStack: 'space000',
      input: {
        size: 'sizing050',
        stylePreset: 'checkboxInput',
        spaceInline: 'space030',
      },
      feedback: {
        size: 'sizing070',
        stylePreset: 'checkboxFeedback',
      },
      label: {
        stylePreset: 'controlLabel',
        typographyPreset: 'utilityBody020',
      },
    },
    medium: {
      spaceStack: 'space000',
      input: {
        size: 'sizing060',
        stylePreset: 'checkboxInput',
        spaceInline: 'space030',
      },
      feedback: {
        size: 'sizing080',
        stylePreset: 'checkboxFeedback',
      },
      label: {
        stylePreset: 'controlLabel',
        typographyPreset: 'utilityBody020',
      },
    },
    large: {
      spaceStack: 'space000',
      input: {
        size: 'sizing070',
        stylePreset: 'checkboxInput',
        spaceInline: 'space040',
      },
      feedback: {
        size: 'sizing090',
        stylePreset: 'checkboxFeedback',
      },
      label: {
        stylePreset: 'controlLabel',
        typographyPreset: 'utilityBody030',
      },
    },
  },
};
