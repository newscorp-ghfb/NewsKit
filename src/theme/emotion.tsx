import React from 'react';
import {
  withTheme as _withTheme,
  useTheme as _useTheme,
  ThemeProvider as EmotionThemeProvider,
} from '@emotion/react';
import {Theme, UncompiledTheme} from './types';
import {compileTheme} from './compiler';
import {themeDiff, CssVariablesContainer, ThemeDiff} from './css-variables';

export const withTheme = (_withTheme as unknown) as <P>(
  comp: React.ComponentType<P & {theme: Theme}>,
) => React.ComponentType<P & {theme?: Theme}>;

export const useTheme = (_useTheme as unknown) as () => Theme;

export interface ThemeProviderProps {
  theme: UncompiledTheme | Theme | ((outerTheme: Theme) => Theme);
  children: React.ReactNode;
  exposeCssVariables?: boolean;
  useThemeCache?: boolean;
}

export const ThemeProvider: React.FC<ThemeProviderProps> = ({
  theme,
  exposeCssVariables = false,
  children,
  ...props
}) => {
  const parentTheme = useTheme();
  const localTheme =
    typeof theme === 'function' ? theme(parentTheme) : compileTheme(theme);

  const diff: ThemeDiff = exposeCssVariables
    ? themeDiff(parentTheme, localTheme)
    : {};

  const doNotAddCssVariables =
    !exposeCssVariables || Object.keys(diff).length === 0;

  return (
    <EmotionThemeProvider {...props} theme={localTheme}>
      {doNotAddCssVariables ? (
        children
      ) : (
        <CssVariablesContainer diff={diff}>{children}</CssVariablesContainer>
      )}
    </EmotionThemeProvider>
  );
};
