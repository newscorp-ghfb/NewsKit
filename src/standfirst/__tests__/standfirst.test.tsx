import {renderToFragmentWithTheme} from '../../test/test-utils';
import {Standfirst} from '..';
import {createTheme} from '../../theme';

const theme = createTheme({
  overrides: {
    stylePresets: {
      linkInline: {
        base: {
          color: '{{colors.interactivePrimary030}}',
          iconColor: '{{colors.interactivePrimary030}}',
          textDecoration: 'underline',
        },
        hover: {
          color: '{{colors.interactivePrimary040}}',
          iconColor: '{{colors.interactivePrimary040}}',
          textDecoration: 'underline',
        },
        active: {
          color: '{{colors.interactivePrimary050}}',
          iconColor: '{{colors.interactivePrimary050}}',
          textDecoration: 'underline',
        },
        visited: {
          color: '{{colors.interactiveVisited010}}',
          iconColor: '{{colors.interactiveVisited010}}',
          textDecoration: 'underline',
        },
        'visited:hover': {
          color: '{{colors.interactiveVisited010}}',
          iconColor: '{{colors.interactiveVisited010}}',
          textDecoration: 'underline',
        },
      },
    },
  },
});

describe('Standfirst', () => {
  test('renders with default styles', () => {
    const fragment = renderToFragmentWithTheme(Standfirst);
    expect(fragment).toMatchSnapshot();
  });

  const tags: Array<['h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'span']> = [
    ['h4'],
    ['span'],
  ];
  test.each(tags)('renders as %s html tag', as => {
    const fragment = renderToFragmentWithTheme(Standfirst, {
      as,
    });
    expect(fragment).toMatchSnapshot();
  });

  test('renders with custom style preset', () => {
    const fragment = renderToFragmentWithTheme(
      Standfirst,
      {
        overrides: {
          styledText: {
            stylePreset: 'linkInline',
          },
        },
      },
      theme,
    );
    expect(fragment).toMatchSnapshot();
  });

  test('renders with custom typography preset', () => {
    const fragment = renderToFragmentWithTheme(Standfirst, {
      overrides: {
        styledText: {
          typographyPreset: 'editorialSubheadline010',
        },
      },
    });
    expect(fragment).toMatchSnapshot();
  });
});
