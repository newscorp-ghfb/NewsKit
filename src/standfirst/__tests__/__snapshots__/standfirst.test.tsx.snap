// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Standfirst renders as h4 html tag 1`] = `
<DocumentFragment>
  .emotion-0 {
  color: #2E2E2E;
}

@media screen and (max-width: 1023px) {
  .emotion-0 {
    font-family: "Bitter",serif;
    font-size: 16px;
    line-height: 18px;
    font-weight: 400;
    letter-spacing: 0;
    padding: 0.5px 0px;
  }

  .emotion-0::before {
    content: '';
    margin-bottom: -0.2055em;
    display: block;
  }

  .emotion-0::after {
    content: '';
    margin-top: -0.2275em;
    display: block;
  }
}

@media screen and (min-width: 1024px) {
  .emotion-0 {
    font-family: "Bitter",serif;
    font-size: 20px;
    line-height: 22.5px;
    font-weight: 400;
    letter-spacing: 0;
    padding: 0.5px 0px;
  }

  .emotion-0::before {
    content: '';
    margin-bottom: -0.2055em;
    display: block;
  }

  .emotion-0::after {
    content: '';
    margin-top: -0.2275em;
    display: block;
  }
}

<h4
    class="emotion-0"
  />
</DocumentFragment>
`;

exports[`Standfirst renders as span html tag 1`] = `
<DocumentFragment>
  .emotion-0 {
  color: #2E2E2E;
  display: inline-block;
}

@media screen and (max-width: 1023px) {
  .emotion-0 {
    font-family: "Bitter",serif;
    font-size: 16px;
    line-height: 18px;
    font-weight: 400;
    letter-spacing: 0;
    padding: 0.5px 0px;
  }

  .emotion-0::before {
    content: '';
    margin-bottom: -0.2055em;
    display: block;
  }

  .emotion-0::after {
    content: '';
    margin-top: -0.2275em;
    display: block;
  }
}

@media screen and (min-width: 1024px) {
  .emotion-0 {
    font-family: "Bitter",serif;
    font-size: 20px;
    line-height: 22.5px;
    font-weight: 400;
    letter-spacing: 0;
    padding: 0.5px 0px;
  }

  .emotion-0::before {
    content: '';
    margin-bottom: -0.2055em;
    display: block;
  }

  .emotion-0::after {
    content: '';
    margin-top: -0.2275em;
    display: block;
  }
}

<span
    class="emotion-0"
  />
</DocumentFragment>
`;

exports[`Standfirst renders with custom style preset 1`] = `
<DocumentFragment>
  .emotion-0 {
  color: #3358CC;
}

@media screen and (max-width: 1023px) {
  .emotion-0 {
    font-family: "Bitter",serif;
    font-size: 16px;
    line-height: 18px;
    font-weight: 400;
    letter-spacing: 0;
    padding: 0.5px 0px;
  }

  .emotion-0::before {
    content: '';
    margin-bottom: -0.2055em;
    display: block;
  }

  .emotion-0::after {
    content: '';
    margin-top: -0.2275em;
    display: block;
  }
}

@media screen and (min-width: 1024px) {
  .emotion-0 {
    font-family: "Bitter",serif;
    font-size: 20px;
    line-height: 22.5px;
    font-weight: 400;
    letter-spacing: 0;
    padding: 0.5px 0px;
  }

  .emotion-0::before {
    content: '';
    margin-bottom: -0.2055em;
    display: block;
  }

  .emotion-0::after {
    content: '';
    margin-top: -0.2275em;
    display: block;
  }
}

<h2
    class="emotion-0"
  />
</DocumentFragment>
`;

exports[`Standfirst renders with custom typography preset 1`] = `
<DocumentFragment>
  .emotion-0 {
  color: #2E2E2E;
  font-family: "Bitter",serif;
  font-size: 16px;
  line-height: 18px;
  font-weight: 400;
  letter-spacing: 0;
  padding: 0.5px 0px;
}

.emotion-0::before {
  content: '';
  margin-bottom: -0.2055em;
  display: block;
}

.emotion-0::after {
  content: '';
  margin-top: -0.2275em;
  display: block;
}

<h2
    class="emotion-0"
  />
</DocumentFragment>
`;

exports[`Standfirst renders with default styles 1`] = `
<DocumentFragment>
  .emotion-0 {
  color: #2E2E2E;
}

@media screen and (max-width: 1023px) {
  .emotion-0 {
    font-family: "Bitter",serif;
    font-size: 16px;
    line-height: 18px;
    font-weight: 400;
    letter-spacing: 0;
    padding: 0.5px 0px;
  }

  .emotion-0::before {
    content: '';
    margin-bottom: -0.2055em;
    display: block;
  }

  .emotion-0::after {
    content: '';
    margin-top: -0.2275em;
    display: block;
  }
}

@media screen and (min-width: 1024px) {
  .emotion-0 {
    font-family: "Bitter",serif;
    font-size: 20px;
    line-height: 22.5px;
    font-weight: 400;
    letter-spacing: 0;
    padding: 0.5px 0px;
  }

  .emotion-0::before {
    content: '';
    margin-bottom: -0.2055em;
    display: block;
  }

  .emotion-0::after {
    content: '';
    margin-top: -0.2275em;
    display: block;
  }
}

<h2
    class="emotion-0"
  />
</DocumentFragment>
`;
